src/Domain/InterfaceComputeTags.hpp

```diff
@@ -35,147 +35,86 @@ namespace Tags {
+/// `Tag` (i.e. Tag::type), with the set of directions being those produced by
+/// `DirectionsTag`. `Tag::function` will be applied separately to the data on
+/// each interface. If some of the compute item's inputs (i.e.
+/// Tag::argument_tags) should be taken from the volume rather than the
+/// interface, they should be specified in the typelist `Tag::volume_tags`.
 ///
-/// \tparam DirectionsTag the item of Directions
-/// \tparam Tag the tag labeling the item
+/// \tparam DirectionsTag the simple tag labeling the set of Directions
+/// \tparam Tag the compute tag to apply on each interface
 template <typename DirectionsTag, typename Tag>
-struct InterfaceCompute : Interface<DirectionsTag, Tag>, db::ComputeTag {
-  using base = Interface<DirectionsTag, Tag>;
+struct InterfaceCompute : Interface<DirectionsTag, typename Tag::base>,
+                          db::ComputeTag {
+  static_assert(db::is_simple_tag_v<DirectionsTag>);
   static_assert(db::is_compute_tag_v<Tag>,
                 "Cannot use a non compute item as an interface compute item.");
-  // Defining name here prevents an ambiguous function call when using base
-  // tags; Both Interface<Dirs, Tag> and Interface<Dirs, Tag::base> will have a
-  // name function and so cannot be disambiguated.
-  static std::string name() noexcept {
-    return "Interface<" + db::tag_name<DirectionsTag>() + ", " +
-           db::tag_name<Tag>() + ">";
-  };
-  using tag = Tag;
+  using base = Interface<DirectionsTag, typename Tag::base>;
+  static constexpr size_t volume_dim = DirectionsTag::volume_dim;
+  using return_type =
+      std::unordered_map<::Direction<volume_dim>, typename Tag::type>;
+
   using forwarded_argument_tags =
       InterfaceHelpers_detail::get_interface_argument_tags<Tag, DirectionsTag>;
   using argument_tags =
       tmpl::push_front<forwarded_argument_tags, DirectionsTag>;
 
-  using return_type = typename Interface_detail::evaluate_compute_item<
-      DirectionsTag, Tag, forwarded_argument_tags>::return_type;
-  template <typename... Ts>
-  static constexpr auto function(Ts&&... ts) noexcept {
-    return Interface_detail::evaluate_compute_item<
-        DirectionsTag, Tag,
-        forwarded_argument_tags>::apply(std::forward<Ts>(ts)...);
+  using volume_tags = get_volume_tags<Tag>;
+  static_assert(
+      tmpl::size<tmpl::list_difference<volume_tags, argument_tags>>::value == 0,
+      "volume_tags contains tags not in argument_tags");
+
+  template <typename... ArgTypes>
+  static constexpr void function(
+      const gsl::not_null<return_type*> result,
+      const std::unordered_set<::Direction<volume_dim>>& directions,
+      ArgTypes&&... args) noexcept {
+    for (const auto& direction : directions) {
+      Interface_detail::evaluate_compute_item<Tag, volume_tags>(
+          make_not_null(&(*result)[direction]), direction,
+          forwarded_argument_tags{}, std::forward<ArgTypes>(args)...);
```

Comment:
`evaluate_compute_item` takes its last arguments as const references, so take `args` as const references here and remove the `std::forward` so it is clear that none of the `args` are being moved out of in multiple iterations of the loop.